app/mailers/application_mailer.rb

```diff
@@ -36,4 +37,37 @@ def generate_unsubscribe_token(id, email_type)
   def use_custom_host
     ActionMailer::Base.default_url_options[:host] = Settings::General.app_domain
   end
+
+  def perform_deliveries?
+    if Rails.env.production?
+      self.perform_deliveries = Settings::SMTP.password.present? ||
+        ENV["SENDGRID_API_KEY"].present?
+    else
+      Rails.configuration.action_mailer.perform_deliveries
+    end
+  end
+
+  protected
+
+  def set_delivery_options
+    self.smtp_settings = if ENV["SENDGRID_API_KEY"].present?
+                           {
+                             address: "smtp.sendgrid.net",
+                             port: 587,
+                             authentication: :plain,
+                             user_name: "apikey",
+                             password: ENV["SENDGRID_API_KEY"],
+                             domain: ENV["APP_DOMAIN"]
+                           }
+                         else
+                           {
+                             address: Settings::SMTP.address,
+                             port: Settings::SMTP.port,
+                             authentication: Settings::SMTP.authentication,
+                             user_name: Settings::SMTP.user_name,
+                             password: Settings::SMTP.password,
+                             domain: Settings::SMTP.domain
+                           }
+                         end
```

Comment:
These could be part of the `Settings::SMTP` class, for example:

```ruby
    class << self
      def settings
        return sendgrid_settings if ENV["SENDGRID_API_KEY"].present?

        keys.map {|k| [k, public_send(k) ]}.to_h
      end

      private

      def sendgrid_settings
        {
          address: "smtp.sendgrid.net",
          port: 587,
          authentication: :plain,
          user_name: "apikey",
          password: ENV["SENDGRID_API_KEY"],
          domain: ENV["APP_DOMAIN"]
        }
      end
    end
```

For example:

```ruby
> Settings::SMTP.settings
=> {"address"=>"smtp.gmail.com", "port"=>25, "authentication"=>nil, "user_name"=>nil, "password"=>nil, "domain"=>nil}
```

So here you can do:

```ruby
self.smtp_settings = Settings::SMTP.settings
```

I would also add some tests to `Settings::SMTP.settings`, to make sure the sendgrid values are unchanged

Reply:
:star_struck: amazing suggestion! thank you!